Stratified Group KFold implementation

marrodion opened the pull request:
#### Reference Issues/PRs
Fixes #13621.

#### What does this implement/fix? Explain your changes.
Implementation of StratifiedGroupKFold based on the PR https://github.com/scikit-learn/scikit-learn/pull/15239 and [kaggle kernel](https://www.kaggle.com/jakubwasikowski/stratified-group-k-fold-cross-validation). 
Implementation considers distribution of labels within the groups without the restriction of each group having only one class.
For trivial cases:
- If number of groups equals to number of samples the implementation follows `StratifiedKFold` as much as possible
- If each group has roughly equal distribution of labels, the behavior is almost identical to `GroupKFold`

Issues with current implementation:
- Due to the sorting of groups when iterating, shuffling happens only for groups with the same label distribution. Not sure if the option to shuffle should be kept at all given that.
- Because the algorithm looks at groups in sequence, it may produce suboptimal splits in terms of stratification (see `.sklearn/model_selection/tests/test_split.py:559` - `test_startified_group_kfold_approximate`)

#### Any other comments?

Given outlined restrictions, I am hesitant whether this should be included into scikit-learn or not. It seems that its usefulness is limited. However, there seem to be a constant interest in this feature and I failed to design a better solution that is not brute-force. Would like to hear any thoughts/ideas on which algorithm might produce better results.


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

hermidalc committed:
Initial implementation

hermidalc committed:
Forgot to add to second __add__ list

hermidalc committed:
Update split method parameter doc

hermidalc committed:
Added example; changed default test_size to 0.1; added to author list

hermidalc committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into stratified-groupshufflesplit

hermidalc committed:
StratifiedGroupKFold impl and other improvements

hermidalc committed:
Add class to __all__ spec

hermidalc committed:
Remove random_state when no shuffle

hermidalc committed:
Tighter formatting

Commit:
Merge branch 'master' into stratified-groupshufflesplit

Commit:
Update the implementation of StratifiedGroupKFold

Commit:
Add StratifiedGroupKFold to __init__

Commit:
Add y checks to StartifiedGroupKFold

Commit:
Raise error if n_splits > max num samples in class

Commit:
Warn if n_splits > mn num samples in class

Commit:
Add SGKfold to general repr test

Commit:
Add SGKFold to 2d_y test case

Commit:
Add SGKfold to value erros test case

Parameters are the same as for StratifiedKFold
to ensure similar behavior given n_groups == n_samples

Commit:
Add SGKFold to StratifiedKFold test cases

The idea is to ensure similar behavior when groups are trivial
(n_groups == n_samples)

Commit:
Add SGKFold to reproducibility test case

Commit:
Add SGKFold to GroupKFold test case

Commit:
Add SGKFold to nested cv test case

Commit:
Add SGKFold to random_state with shuffle=False test case

Commit:
Add SGKFold to constant splits test case

Commit:
Fix repr test case

Commit:
Fix formatting issues

Commit:
Add samples to a fold with least num samples

Required to produce balanced size folds when the distribution of y is
more or less the same

Commit:
Remove GroupShuffleSplit impl

Commit:
Add notes to StratifiedGroupKFold

Commit:
Fix doctest

Commit:
Added stratified group kfold tests

Commit:
Better variable naming

Commit:
Add section to documentation

Commit:
Merge branch 'master' into stratified-group-kfold

Commit:
Remove leftover StratifiedGroupShuffleSplit import

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -732,6 +735,166 @@ def split(self, X, y, groups=None):
+            raise ValueError("n_splits=%d cannot be greater than the"
+                             " number of members in each class."
+                             % (self.n_splits))
+        min_groups = np.min(y_cnt)
```
this seems to be the wrong name ... maybe `max_splits` or `n_smallest_class`.

marrodion replied:
Thanks, changed to `n_smallest_class`

jnothman commented on doc/modules/cross_validation.rst:
```diff
@@ -648,6 +648,39 @@ Here is a visualization of the cross-validation behavior.
+Implementation notes:
+
+- This implementation iterates over groups sorted by standard deviation of
+  labels. That means that when shuffle=True only groups with the same labels
```
"labels" is confusing because both groups and classes have labels. Use "classes".

marrodion replied:
Thank you, changed

jnothman replied:
Not changed here, though?

marrodion replied:
My apologies and thanks for pointing out.
Should be fixed now.

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -732,6 +735,166 @@ def split(self, X, y, groups=None):
+                         random_state=random_state)
+
+    # Implementation based on this kaggle kernel:
+    # https://www.kaggle.com/jakubwasikowski/stratified-group-k-fold-cross-validation
```
This is released under Apache 2.0 license. Please make sure that its conditions have been fulfilled.

marrodion replied:
I'm afraid I need some advice here..

As far as I understood, Apache 2.0 has three main conditions:
- Add a list of changes made to the original code. I've added it to the docstring of the class. Is this a right place or should it be in the function body as a block comment?
- Provide a copy of copyright and notice files. There are no copyright or notice files in Kaggle kernel, not sure what if that's ok to leave as is or I should add copyright from [Apache license example](http://www.apache.org/licenses/LICENSE-2.0#apply). If the latter, should this copyright be issued to a Kaggle user or to Kaggle?
- Provide a copy of a license. Is link to [Apache license](http://www.apache.org/licenses/LICENSE-2.0) ok or should it be put in the code as a file? If the latter, what is the best place to put it in?

My apologies, I know that's a lot of questions, I just don't have a lot of experience with open source licensing.

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -732,6 +735,166 @@ def split(self, X, y, groups=None):
+            warnings.warn(("The least populated class in y has only %d"
+                           " members, which is less than n_splits=%d."
+                           % (min_groups, self.n_splits)), UserWarning)
+        labels_num = np.max(y_inv) + 1
```
```suggestion
        labels_num = len(y_cnt)
```

marrodion replied:
Thanks, totally missed that. Should be fixed now.

jnothman reviewed (commented):
Thank you for the PR. I look forward to progressing this.

I think this algorithm requires some description. Here's my attempt.

The algorithm greedily assigns each group to one of `n_splits` test sets, choosing the test set that minimises the variance in class distribution across test sets. Group assignment proceeds from groups with highest to lowest variance in class frequency, i.e. large groups peaked on one or few classes are assigned first.

I'm fairly happy with this so far, but am concerned that we choose a reasonable greedy heuristic and a reasonable assignment order; and that we do our best to test correctness and appropriateness to a few common use cases.

marrodion committed:
Merge remote-tracking branch 'upstream/master' into stratified-group-kfold

marrodion committed:
Add changelist and reference to original kernel

marrodion committed:
Better naming for least populated class check

marrodion committed:
Better expression for number of labels

marrodion committed:
Remove use of Counter

We already have this data in output of np.unique

marrodion committed:
Add tests for homogeneous groups

marrodion committed:
Add StratifiedGroupKFold test against GroupKFold

marrodion committed:
Add changes to changelist in docstring

marrodion committed:
Add StratifiedGroupKFold to classes.rst

marrodion committed:
Fix description of StratifiedGroupKFold

marrodion commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -509,6 +538,42 @@ def test_kfold_can_detect_dependent_samples_on_digits():  # see #2372
+        assert_allclose(split_dist, expect_dist, atol=0.001)
+        test_sizes.append(len(test))
+    assert np.ptp(test_sizes) <= 1
+
```
Added test case for that, see ./sklearn/model_selection/tests/test_split.py:588

marrodion commented:
Thank you so much for the review. I fixed most of the issues you've outlined, except for licensing one.

> we choose a reasonable greedy heuristic and a reasonable assignment order; and that we do our best to test correctness and appropriateness to a few common use cases.

To verify that, the best I could come up with is to add a test where `StratifiedGroupKFold` is compared to a regular `GroupKFold` in terms of stratification quality on several random data samples. See `sklearn/model_selection/tests/test_split.py:598` Please, let me know if that makes sense or should I implement something else.

wang701 commented:
Hello @marrodion, sorry to hijack this PR; from #13621 it seems like you are actively working on this and that's why I want to ask the question here.

In the notes of `StratifiedGroupKFold` in [_split.py](https://github.com/scikit-learn/scikit-learn/blob/2580a814fd54df12ca0d3a8e9c67818ece224d34/sklearn/model_selection/_split.py), the third bullet point states:

> Stratify based on samples as much as possible while keeping non-overlapping groups constraint. That means that in some cases when there is a small number of groups containing a large number of samples the stratification will not be possible and the behavior will be close to GroupKFold.

I do have a dataset just like that (i.e., "a small number of groups containing a large number of samples") and the resultant splits using @hermidalc code in [comments](https://github.com/scikit-learn/scikit-learn/issues/13621#issuecomment-656094573) are really close to GroupKFold. Why does it happen? I am a complete novice in the implementation of sklearn so any pointer is appreciated.

marrodion commented:
Hello @wang701,

> I do have a dataset just like that (i.e., "a small number of groups containing a large number of samples") and the resultant splits using @hermidalc code in comments are really close to GroupKFold. Why does it happen?

The idea of the algorithm is to assign groups to folds in a way that would keep distribution of classes in each fold similar. If you have a small number of groups with different distribution of classes, then there is no assignment you could do that would make distributions similar *and* keep each group in a single fold.

marrodion commented:
Hello @jnothman,

Is there anything I could/should do to progress this PR?
Should I do the issues left (license) the way I think they should be done and then fix after your review?

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -3,13 +3,15 @@
+#         Gael Varoquaux <gael.varoquaux@normalesup.org>
 #         Olivier Grisel <olivier.grisel@ensta.org>
 #         Raghav RV <rvraghav93@gmail.com>
+#         Leandro Hermida <hermidal@cs.umd.edu>
```
I'm confused. Is the implementation here the one by Leandro?

marrodion replied:
It is a continuation of his [PR](https://github.com/scikit-learn/scikit-learn/pull/15239), so I left this as is.

hermidalc replied:
I'm glad my initial implementation was of use, please also add yourself and others who contributed!

hermidalc replied:
Actually that early implementation in #15239 I haven't used anymore, this is the other implementation I wrote up and use all the time https://github.com/scikit-learn/scikit-learn/issues/13621#issuecomment-600894432.  If you built from this thank you for including my name.

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -732,6 +735,176 @@ def split(self, X, y, groups=None):
+    * Implementation is based on this kaggle kernel:
+      https://www.kaggle.com/jakubwasikowski/stratified-group-k-fold-cross-validation
+
+      Changelist:
```
This list of changes is to comply with license, and does not need to be visible in the code. Rather, the license needs to be copied into a _comment_, saying that this class is subject to the Apache 2.0 license, citing its source, and noting the changes.

jnothman reviewed (commented):
Sorry I'm not yet reviewing substance here!

cmarmo commented:
Hi @marrodion, thanks for your patience! Are you still interested in working on this? If so, do you mind addressing the comment about the license? Also, a new commit will trigger the builds and make the review easier (especially for documentation). Thanks for your collaboration.

marrodion committed:
Merge branch 'main' into stratified-group-kfold

marrodion committed:
Move license notice out of docstring

marrodion committed:
Disambiguate labels to classes in doc

marrodion commented:
Hello @cmarmo, thank you for your comment.
I've made the change:
- extracted license notice from docstring to comment
- fixed documentation issue pointed out by @jnothman 

I'd like to finish this PR, so will try to be more reactive to notifications on github - I somehow missed the previous comments.

If there is anything else I need to do - please, let me know.

marrodion committed:
Merge remote-tracking branch 'upstream/main' into stratified-group-kfold

marrodion committed:
Merge remote-tracking branch 'upstream/main' into stratified-group-kfold

marrodion committed:
Add changelog entry

marrodion commented:
@cmarmo @jnothman 
Sorry to bother, but I think your comments were fixed and now this branch does't have conflicts/builds properly.
Is there anything else I need to do to progress with this PR?

cmarmo commented:
Hi @marrodion thanks for your patience : I have added your PR to the 1.0 milestone. Let's see if this can speed up things a bit.

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -732,6 +735,177 @@ def split(self, X, y, groups=None):
+            min_samples_in_fold = np.inf
+            for i in range(self.n_splits):
+                y_counts_per_fold[i] += y_counts
+                std_per_label = []
+                for label in range(labels_num):
+                    std_per_label.append(np.std(
+                        [y_counts_per_fold[j][label] / y_cnt[label]
+                         for j in range(self.n_splits)]))
```
Since `y_counts_per_fold` is now an array, I think this is equivalent:
```suggestion
                # Summarise the distribution over classes in each proposed fold
                std_per_label = (y_counts_per_fold / y_cnt.reshape(1, -1)).std(axis=0)
```
But let's call this `std_per_class`?

jnothman replied:
Just to check, is this intentionally normalising over the full `y_cnt`, rather than the `y_cnt` allocated so far (i.e. `y_counts_per_fold.sum(axis=0)`)?

marrodion replied:
> Just to check, is this intentionally normalising over the full y_cnt, rather than the y_cnt allocated so far (i.e. y_counts_per_fold.sum(axis=0))?

Yes, for two reasons:
- to have a single normalization constant for all groups/folds
- to avoid division by zero for groups with a single class

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -732,6 +735,177 @@ def split(self, X, y, groups=None):
+                         for j in range(self.n_splits)]))
+                y_counts_per_fold[i] -= y_counts
+                fold_eval = np.mean(std_per_label)
+                samples_in_fold = np.sum(y_counts_per_fold[i])
```
Just to check: is this intentionally after removing `y_counts` from `y_counts_per_fold`?

marrodion replied:
Yes, the idea is to find the fold with least number of samples before the addition of the current group (which happens several lines earlier)

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -732,6 +735,177 @@ def split(self, X, y, groups=None):
+
+        for group, y_counts in sorted(groups_and_y_counts,
+                                      key=lambda x: -np.std(x[1])):
+            best_fold = None
```
It's tempting to pull this out into a separate method
```
@staticmethod
def _find_best_fold(y_counts, y_counts_per_fold, group_y_counts):
    ...
    return best_fold        
```

jnothman reviewed (commented):
Just looking over the code and docs (not tests). It is hard to read at the moment, in part due to the mixture of Python and numpy idioms, and in part due to confusing variable names. I hope these comments help improve readability.

marrodion committed:
Fix changelog author entry

marrodion committed:
Fix StratifiedGroupKFold docstring

marrodion committed:
Better variable names

marrodion committed:
Remove defaultdict in favor of numpy indexing

marrodion committed:
Make use of numpy broadcasting instead of for loop

marrodion committed:
Encode groups and use arrays instead of dicts

marrodion committed:
Use numpy sort instead of python

marrodion committed:
Clarify shuffling behavior of StratifiedGroupKF in docs

marrodion committed:
Switch name from label_idx to class_idx

marrodion committed:
Extracted best_fold search into a separate method

marrodion committed:
Merge remote-tracking branch 'upstream/main' into stratified-group-kfold

marrodion commented on doc/modules/cross_validation.rst:
```diff
@@ -650,6 +650,44 @@ Here is a visualization of the cross-validation behavior.
+Implementation notes:
+
+- This implementation iterates over groups sorted by standard deviation of
+  classes. That means that when shuffle=True only groups with the same classes
```
Thank you, fixed

marrodion commented on doc/modules/cross_validation.rst:
```diff
@@ -650,6 +650,44 @@ Here is a visualization of the cross-validation behavior.
+
+- This implementation iterates over groups sorted by standard deviation of
+  classes. That means that when shuffle=True only groups with the same classes
+  distribution will be shuffled.
```
Yes, you are right. I've changed the description, detailing a bit more what happens if the user specifies `shuffle=True`. Hope that makes the description better.

marrodion commented on doc/modules/cross_validation.rst:
```diff
@@ -650,6 +650,44 @@ Here is a visualization of the cross-validation behavior.
+  classes are assigned first.
+- This split is suboptimal in a sense that it might produce imbalanced splits
+  even if perfect stratification is possible. If you have relatively close
+  distribution of labels within groups, using :class:`GroupKFold` is better.
```
Thank you, fixed

marrodion commented on doc/whats_new/v1.0.rst:
```diff
@@ -169,6 +169,17 @@ Changelog
+  :class:`model_selection.StratifiedKFold` and `model_selection.GroupKFold`,
+  providing an ability to split data preserving the distribution of classes in
+  each split while keeping each group within a single split.
+  :pr:`18649` by `Leandro Hermida <hermidalc>` `Rodion Martynov <marrodion>`.
```
Thanks for noticing, fixed

marrodion commented on sklearn/model_selection/_split.py:
```diff
@@ -732,6 +735,177 @@ def split(self, X, y, groups=None):
+            warnings.warn(("The least populated class in y has only %d"
+                           " members, which is less than n_splits=%d."
+                           % (n_smallest_class, self.n_splits)), UserWarning)
+        labels_num = len(y_cnt)
```
Absolutely, thanks for suggestion
Fixed

marrodion commented on sklearn/model_selection/_split.py:
```diff
@@ -732,6 +735,177 @@ def split(self, X, y, groups=None):
+        for label, group in zip(y_inv, groups):
+            y_counts_per_group[group][label] += 1
+
+        y_counts_per_fold = defaultdict(lambda: np.zeros(labels_num))
```
Haven't thought about that, fixed

marrodion committed:
Remove accidentally leftover comment

marrodion committed:
Fix np.sort keyword to support numpy < 1.15

marrodion commented:
Thank you so much for a thorough review, @jnothman!

The issues you've outlined seem to be fixed now.
I've also tried to switch from pure python to numpy wherever I could - hope that helps with readability.

agramfort commented on sklearn/model_selection/_split.py:
```diff
@@ -732,6 +735,190 @@ def split(self, X, y, groups=None):
+    The implementation is designed to:
+
+    * Mimic the behavior of StratifiedKFold as much as possible for trivial
+      groups (e.g. when each group contain only one sample).
```
```suggestion
      groups (e.g. when each group contains only one sample).
```

agramfort reviewed (commented):
can you also update https://scikit-learn.org/stable/auto_examples/model_selection/plot_cv_indices.html to share a visual intuition of what's going on?

thx a lot @marrodion

marrodion committed:
Merge remote-tracking branch 'upstream/main' into stratified-group-kfold

marrodion committed:
Fix typo in docstring

marrodion committed:
Add StratifiedGroupKFold to visualization doc

marrodion committed:
Merge remote-tracking branch 'upstream/main' into stratified-group-kfold

marrodion commented:
@agramfort Thank you for your comment,

I've updated the visualization, however, the data in the example makes current algorithm perform identical to `GroupKFold`. Not sure if I should add an additional plot with something like an example from https://github.com/scikit-learn/scikit-learn/issues/13621#issuecomment-557802602. What would you suggest?

agramfort commented:
yes I would add something to the example so we can visually explain what it does

marrodion committed:
Add visualization for uneven group as an example

marrodion commented:
I've added it as an example comparing `StratifiedKFold`, `GroupKFold` and `StratifiedGroupKFold` for uneven groups. That should emphasize the difference between `GroupKFold` and `StratifiedGroupKFold`, that is not visible in the visualization for all CV methods.

agramfort reviewed (approved):
thx @marrodion !

jnothman commented:
this isn't rendering right: images are in the wrong sections. https://134931-843222-gh.circle-artifacts.com/0/doc/modules/cross_validation.html

jnothman commented:
I think the change to the example affected the numbering of images exported and used in the user guide.

jnothman commented:
The image is actually a very helpful demonstration of how it differs. Thanks both!

marrodion committed:
Fix image numbers to match updated example

marrodion commented:
> this isn't rendering right: images are in the wrong sections.

Seems to be fixed now: https://134972-843222-gh.circle-artifacts.com/0/doc/modules/cross_validation.html

jnothman commented on doc/modules/cross_validation.rst:
```diff
@@ -645,11 +645,56 @@ size due to the imbalance in the data.
+- This split is suboptimal in a sense that it might produce imbalanced splits
+  even if perfect stratification is possible. If you have relatively close
+  distribution of classes in each group, using :class:`GroupKFold` is better.
+
```
Insert the visualisation here?

marrodion replied:
Done, thank you.
Wasn't sure if needed, not every CV has a visualization in this documentation page.

jnothman replied:
Perhaps not needed, but helpful. Happy for you to make the docs more consistent in another pr! ;)

jnothman commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -318,20 +344,22 @@ def test_stratified_kfold_no_shuffle():
     y = np.array([4] * int(0.10 * n_samples) +
                  [0] * int(0.89 * n_samples) +
                  [1] * int(0.01 * n_samples))
+    groups = np.arange(len(y))
```
```suggestion
    groups = np.arange(len(y))  # ensure perfect stratification with StratifiedGroupKFold
```

jnothman commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -340,20 +368,22 @@ def test_stratified_kfold_ratios(k, shuffle):
                  [0] * int(0.89 * n_samples) +
                  [1] * int(0.01 * n_samples))
     X = np.ones(len(y))
+    groups = np.arange(len(y))
```
```suggestion
    groups = np.arange(len(y))  # ensure perfect stratification with StratifiedGroupKFold
```

jnothman commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -372,17 +402,19 @@ def test_kfold_balance():
     # Repeat with shuffling turned off and on
     X = np.ones(17)
     y = [0] * 3 + [1] * 14
+    groups = np.arange(len(y))
```
```suggestion
    groups = np.arange(len(y))  # ensure perfect stratification with StratifiedGroupKFold
```

jnothman reviewed (approved):
This is really nice work. I love the tests. Well done.

marrodion committed:
Add author

marrodion committed:
Add SGKF visualization to docs

marrodion committed:
Add comments for groups in stratified CV tests

jnothman commented:
Thank you @marrodion @hermidalc! It's nice to finally have a solution for this case!!

jakubwasikowski commented:
Wow, it's great to see my kernel could be helpful, and I'm happy that the method is finally in scikit-learn. Thanks @marrodion and @hermidalc for doing this 🍻
